wsd/ClientRequestDispatcher.cpp

```diff
@@ -370,6 +372,114 @@ getConvertToBrokerImplementation(const std::string& requestType, const std::stri
     return nullptr;
 }
 
+class ConvertToAddressResolver : public std::enable_shared_from_this<ConvertToAddressResolver>
+{
+    std::shared_ptr<ConvertToAddressResolver> _selfLifecycle;
+    std::queue<std::string> _addressesToResolve;
+    ClientRequestDispatcher::AsyncFn _asyncCb;
+    bool _allow;
+
+public:
+
+    ConvertToAddressResolver(std::queue<std::string> addressesToResolve, ClientRequestDispatcher::AsyncFn asyncCb)
+        : _addressesToResolve(std::move(addressesToResolve))
+        , _asyncCb(asyncCb)
+        , _allow(true)
+    {
+    }
+
+    void testHostName(const std::string& hostToCheck)
+    {
+        _allow &= HostUtil::allowedWopiHost(hostToCheck);
+    }
+
+    // synchronous case
+    bool syncProcess()
+    {
+        assert(!_asyncCb);
+        while (!_addressesToResolve.empty())
+        {
+            const std::string& addressToCheck = _addressesToResolve.front();
+
+            try
+            {
+                std::string resolvedHostName = net::canonicalHostName(addressToCheck);
+                testHostName(resolvedHostName);
+            }
+            catch (const Poco::Exception& exc)
+            {
+                LOG_ERR_S("net::canonicalHostName(\"" << addressToCheck
+                                                      << "\") failed: " << exc.displayText());
+                // We can't find out the hostname, and it already failed the IP check
+                _allow = false;
+            }
+
+            if (_allow)
+            {
+                LOG_INF_S("convert-to: Requesting address is allowed: " << addressToCheck);
+            }
+            else
+            {
+                LOG_WRN_S("convert-to: Requesting address is denied: " << addressToCheck);
+                break;
+            }
+
+            _addressesToResolve.pop();
+        }
+        return _allow;
+    }
+
+    // asynchronous case
+    void startAsyncProcessing()
+    {
+        assert(_asyncCb);
+        _selfLifecycle = shared_from_this();
+        dispatchNextLookup();
+    }
+
+    void dispatchNextLookup()
+    {
+        net::AsyncDNS::DNSThreadFn pushHostnameResolvedToPoll = [this](const std::string& hostname,
+                                                                       const std::string& exception) {
+            COOLWSD::getWebServerPoll()->addCallback([this, hostname, exception]() {
```

Comment:
Great - glad we push into the right thread at the end =) I wonder if API-wise we might want to take a SocketPoll & - as well as a function - so it's clear in the API that we will always want a callback to happen in the right thread and it's not possible to re-use the async-DNS stuff outside that safety-net ;-) but - fine for now I guess.